cellprofiler/gui/figure.py

```diff
@@ -1173,59 +1142,265 @@ def show_hist(evt):
             )
             fig.figure.canvas.draw()
 
-        def change_contrast(evt):
-            """Callback for Image contrast menu items"""
-            axes = self.subplot(x, y)
-            if evt.Id == MENU_CONTRAST_RAW:
-                params["normalize"] = False
-            elif evt.Id == MENU_CONTRAST_NORMALIZED:
-                params["normalize"] = True
-            elif evt.Id == MENU_CONTRAST_LOG:
-                params["normalize"] = "log"
-            elif evt.Id == MENU_CONTRAST_GAMMA:
-                params["normalize"] = "gamma"
-            for artist in axes.artists:
-                if isinstance(artist, cellprofiler.gui.artist.CPImageArtist):
-                    artist.kwargs["normalize"] = params["normalize"]
-                    self.figure.canvas.draw()
-                    return
+        def open_contrast_dialog(evt):
+            nonlocal params
+            orig_params = params.copy()
+            id_dict = {
+                False: "Raw",
+                True: "Normalised",
+                "log": "Log Normalised",
+                "gamma": "Gamma Normalised"
+            }
+            if params["normalize"]:
+                maxval = 1
+                minval = 0
             else:
-                refresh_figure()
+                imgmax = orig_params["vmax"]
+                imgmin = orig_params["vmin"]
+                maxval = max(1, imgmax)
+                minval = min(0, imgmin)
+            start_min = int((params["vmin"]/minval)*255 if minval != 0 else 0)
+            start_max = int((params["vmax"]/maxval)*255)
+            axes = self.subplot(x, y)
+            background = self.figure.canvas.copy_from_bbox(axes.bbox)
+            size = self.images[(x, y)].shape[:2]
+            axesdata = axes.plot([0, 0], list(size), "k")[0]
 
-        def adjust_gamma(evt):
-            dlg = wx.TextEntryDialog(self, "Normalization factor", "Adjust gamma")
-            dlg.SetValue(get_normalization_factor())
-            if dlg.ShowModal() == wx.ID_OK:
-                params["normalize_args"] = {"gamma": float(dlg.GetValue())}
-            dlg.Destroy()
+            with wx.Dialog(self, title="Adjust contrast", size=wx.Size(250, 300)) as dlg:
```

Comment:
Would you mind extracting this into a seperate Python module and class?

Reply:
Note for the future: We decided it'd be best to do this as part of a larger refactor in which we'd split figure.py into a more sensible structure.